gcp/api/server.py

```diff
@@ -111,29 +111,45 @@ def QueryAffectedBatch(self, request, context):
   @ndb_context
   def DetermineVersion(self, request, context):
     """Determine the version of the provided hashes."""
-    if not request.query.name:
-      context.abort(grpc.StatusCode.NOT_IMPLEMENTED,
-                    'Querying only by file hash is not implemented yet.')
-      return None
-    return get_version_by_name(request.query).result()
+    return determine_version(request.query, context).result()
 
 
 @ndb.tasklet
-def get_version_by_name(
-    version_query: osv_service_v1_pb2.VersionQuery) -> ndb.Future:
-  """Identifies the version based on the provided name."""
-  query = osv.RepoIndex.query(osv.RepoIndex.name == version_query.name)
-  it = query.iter()
+def determine_version(version_query: osv_service_v1_pb2.VersionQuery,
+                      context: grpc.ServicerContext) -> ndb.Future:
+  """Identify fitting commits based on a subset of hashes"""
+  hashes = []
+  for i in range(min(20, len(version_query.file_hashes))):
```

Comment:
can you put 20 into a constant? 

also, this can be simpler as:

```python
hashes = [f.hash in f in version_query.file_hashes[:_MAX_HASHES_TO_TRY]]
```

Or we can use `random.sample` as well: https://docs.python.org/3/library/random.html#random.sample instead of just taking the first 20.

Reply:
Sure, thanks!

I played around with random sampling. It doesn't seem like it is improving the time needed. However it strongly depends on the ordering of hashes I assume and with some of the other changes could be an improvement. However,  it might have negative effects when picking common 3p deps adding a lot of overlap.

It looks like it provides a good balance. Since always taking the first X hashes can bring problems too.

Everything below is for protobuf v21.0:

First 20 hashes:
```
zsh grpc_cmd.sh  0.01s user 0.03s system 0% cpu 12.126 total
zsh grpc_cmd.sh  0.02s user 0.03s system 0% cpu 11.516 total
zsh grpc_cmd.sh  0.03s user 0.01s system 0% cpu 10.613 total
zsh grpc_cmd.sh  0.02s user 0.02s system 0% cpu 11.051 total
zsh grpc_cmd.sh  0.02s user 0.03s system 0% cpu 10.430 total
zsh grpc_cmd.sh  0.00s user 0.05s system 0% cpu 10.405 total
zsh grpc_cmd.sh  0.01s user 0.05s system 0% cpu 10.534 total
zsh grpc_cmd.sh  0.02s user 0.02s system 0% cpu 10.765 total
zsh grpc_cmd.sh  0.02s user 0.02s system 0% cpu 10.961 total
zsh grpc_cmd.sh  0.01s user 0.04s system 0% cpu 10.784 total
```

100 random hashes:
```
zsh grpc_cmd.sh  0.01s user 0.03s system 0% cpu 7.781 total
zsh grpc_cmd.sh  0.01s user 0.04s system 0% cpu 23.819 total
zsh grpc_cmd.sh  0.02s user 0.04s system 1% cpu 5.258 total
zsh grpc_cmd.sh  0.04s user 0.01s system 0% cpu 24.728 total
zsh grpc_cmd.sh  0.03s user 0.02s system 0% cpu 5.334 total
zsh grpc_cmd.sh  0.01s user 0.04s system 0% cpu 29.863 total
zsh grpc_cmd.sh  0.02s user 0.03s system 0% cpu 29.197 total
zsh grpc_cmd.sh  0.03s user 0.03s system 0% cpu 28.342 total
zsh grpc_cmd.sh  0.01s user 0.04s system 1% cpu 5.052 total
zsh grpc_cmd.sh  0.03s user 0.02s system 1% cpu 5.147 total
```

50 random hashes:
```
zsh grpc_cmd.sh  0.02s user 0.03s system 0% cpu 26.184 total
zsh grpc_cmd.sh  0.03s user 0.02s system 0% cpu 21.110 total
zsh grpc_cmd.sh  0.01s user 0.05s system 1% cpu 4.614 total
zsh grpc_cmd.sh  0.01s user 0.05s system 1% cpu 4.447 total
zsh grpc_cmd.sh  0.03s user 0.03s system 1% cpu 4.237 total
zsh grpc_cmd.sh  0.03s user 0.02s system 0% cpu 24.868 total
zsh grpc_cmd.sh  0.03s user 0.02s system 1% cpu 3.625 total
zsh grpc_cmd.sh  0.01s user 0.05s system 0% cpu 24.234 total
zsh grpc_cmd.sh  0.02s user 0.03s system 0% cpu 21.577 total
zsh grpc_cmd.sh  0.02s user 0.03s system 0% cpu 22.538 total
```